op-node/rollup/status/l1_head_buffer.go

```diff
@@ -0,0 +1,67 @@
+package status
+
+import (
+	"sync"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+)
+
+// l1HeadBuffer is a thread-safe cache for L1 block references, which contains a series blocks with a valid chain of parent hashes.
+type l1HeadBuffer struct {
+	rb             *ringbuffer[eth.L1BlockRef]
+	minBlockNumber uint64
+	mu             sync.RWMutex
+}
+
+func newL1HeadBuffer(size int) *l1HeadBuffer {
+	return &l1HeadBuffer{rb: newRingBuffer[eth.L1BlockRef](size)}
+}
+
+// Get returns the L1 block reference for the given block number, if it exists in the cache.
+func (lhb *l1HeadBuffer) Get(num uint64) (eth.L1BlockRef, bool) {
+	lhb.mu.RLock()
+	defer lhb.mu.RUnlock()
+
+	return lhb.get(num)
+}
+
+func (lhb *l1HeadBuffer) get(num uint64) (eth.L1BlockRef, bool) {
+	return lhb.rb.Get(int(num - lhb.minBlockNumber))
+}
+
+// Insert inserts a new L1 block reference into the cache, and removes any entries that are invalidated by a reorg.
+// If the parent hash of the new head doesn't match the hash of the previous head, all entries after the new head are removed
+// as the chain cannot be validated.
+func (lhb *l1HeadBuffer) Insert(l1Head eth.L1BlockRef) {
+	lhb.mu.Lock()
+	defer lhb.mu.Unlock()
+
+	if ref, ok := lhb.get(l1Head.Number - 1); ok && ref.Hash == l1Head.ParentHash {
+		// Parent hash is found, so we can safely add the new head to the cache.
+
+		// Check to see if the new L1 head is already in the cache.
+		// If the hash doesn't match the one in the cache, we have a reorg and need to remove all entries after the new head.
+		if ref, ok := lhb.get(l1Head.Number); ok {
+			if ref.Hash != l1Head.Hash {
+				// Reorg detected, invalidate all entries after the new head.
+				for {
+					ref, ok := lhb.rb.Pop()
+					if !ok || ref.Number == l1Head.Number {
+						break
+					}
+				}
```

Comment:
This leaves open the possibility that the head rewinds on the same chain for some reason (not a normal event but is possible if the L1 is a proxy and its backend changes) and we would then return a cached block that's actually past the current chain head. We don't have any signal yet that those blocks aren't canonical so I think that's ok and would save us from resetting the cache if the head then skips forward again suddenly.

In practice the `ConfDepth` restricts requests to ensure they're prior to the l1 head as well. We could be really defensive and apply that restriction here as well but probably don't need to.

Reply:
This was an open question I had as well - if we receive heads `1, 2, 3, 2`, is it better to erase `3` or ignore the second `2`? I was leaning towards the latter (the current implementation), as this is least likely to be cache-destructive while also not presenting contradictory history.

Reply:
Looking at it again, the `L1Fetcher` does get used in some places that aren't wrapped by the `conf_depth` like the `EngineResetDeriver`. That would be a particularly bad place to be able to get blocks from the L1Fetcher that we then can't get the other data from because they're actually after the node's head.

So I think it's probably worth being extra conservative here and just ensuring that `Get` doesn't use cached values that are after the current `lhb.l1Head`.  Then we still get the benefit of preserving the cache if it advances down the same chain again but will never return a cached value we would have removed here if we did truncate back to the new head.

Reply:
Thanks for the input. I'm not looking to cause any chain splits or stalls, so I added the rewind capability :)

Reply:
Sometimes its better to be safe than paged and sorry...  Looks good. Thanks.